File: backend/app/schemas/interview/interview.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class InterviewStartRequest(BaseModel):
    mode: str = Field(..., min_length=2)
    role: str = Field(..., min_length=2)
    difficulty: str = Field(..., min_length=2)
    language: str = "en"
    max_questions: int = Field(default=5, ge=3, le=10)


class InterviewAnswerRequest(BaseModel):
    answer: str = Field(..., min_length=1)


class InterviewMessageResponse(BaseModel):
    id: Optional[int] = None
    session_id: int
    role: str
    content: str
    question_index: Optional[int] = None
    message_type: str = "message"
    meta: Dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class InterviewSessionResponse(BaseModel):
    id: int
    user_id: int

    mode: str
    role: str
    difficulty: str
    language: str

    status: str
    current_question_index: int
    max_questions: int

    resume_context: Dict[str, Any] = Field(default_factory=dict)
    session_context: Dict[str, Any] = Field(default_factory=dict)

    started_at: Optional[datetime] = None
    ended_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class InterviewInsight(BaseModel):
    title: str
    description: str
    impact: str = "medium"


class InterviewResultResponse(BaseModel):
    id: Optional[int] = None
    session_id: int
    user_id: int

    mode: str
    role: str
    difficulty: str
    language: str

    overall_score: int = 0
    confidence_score: int = 0
    communication_score: int = 0
    structure_score: int = 0
    specificity_score: int = 0

    recruiter_engagement: Optional[str] = None
    filler_words: Optional[str] = None
    estimated_confidence: Optional[str] = None

    strengths: List[str] = Field(default_factory=list)
    weaknesses: List[str] = Field(default_factory=list)
    recruiter_insights: List[InterviewInsight] = Field(default_factory=list)
    coaching_tips: List[str] = Field(default_factory=list)

    raw_evaluation: Dict[str, Any] = Field(default_factory=dict)

    created_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class InterviewStartResponse(BaseModel):
    session: InterviewSessionResponse
    first_message: InterviewMessageResponse


class InterviewAnswerResponse(BaseModel):
    session: InterviewSessionResponse
    user_message: InterviewMessageResponse
    ai_message: Optional[InterviewMessageResponse] = None
    result: Optional[InterviewResultResponse] = None


class InterviewSessionDetailResponse(BaseModel):
    session: InterviewSessionResponse
    messages: List[InterviewMessageResponse] = Field(default_factory=list)
    result: Optional[InterviewResultResponse] = None


class InterviewStatsBucket(BaseModel):
    difficulty: str
    sessions: int = 0
    average_score: int = 0
    best_score: int = 0


class InterviewStatsResponse(BaseModel):
    total_sessions: int = 0
    average_score: int = 0
    best_score: int = 0
    by_difficulty: List[InterviewStatsBucket] = Field(default_factory=list)
    recent_results: List[InterviewResultResponse] = Field(default_factory=list)

File: frontend/src/features/interview/types/interview.ts
export type InterviewMode =
  | "behavioral"
  | "technical"
  | "system-design";

export type InterviewDifficulty =
  | "Junior"
  | "Mid"
  | "Senior";

export type InterviewLanguage =
  | "en"
  | "de";

export type InterviewInsightImpact =
  | "low"
  | "medium"
  | "high";

export type InterviewStatus =
  | "active"
  | "completed";

export interface InterviewStartRequest {
  mode: string;
  role: string;
  difficulty: string;
  language: string;
  max_questions: number;
}

export interface InterviewAnswerRequest {
  answer: string;
}

export interface InterviewMessage {
  id?: number;

  session_id: number;

  role: string;
  content: string;

  question_index?: number | null;

  message_type: string;

  meta: Record<string, unknown>;

  created_at?: string | null;
}

export interface InterviewSession {
  id: number;

  user_id: number;

  mode: string;
  role: string;
  difficulty: string;
  language: string;

  status: InterviewStatus;

  current_question_index: number;
  max_questions: number;

  resume_context: Record<string, unknown>;
  session_context: Record<string, unknown>;

  started_at?: string | null;
  ended_at?: string | null;

  created_at?: string | null;
  updated_at?: string | null;
}

export interface InterviewInsight {
  title: string;
  description: string;
  impact: InterviewInsightImpact;
}

export interface InterviewResult {
  id?: number;

  session_id: number;
  user_id: number;

  mode: string;
  role: string;
  difficulty: string;
  language: string;

  overall_score: number;

  confidence_score: number;
  communication_score: number;
  structure_score: number;
  specificity_score: number;

  recruiter_engagement?: string | null;
  filler_words?: string | null;
  estimated_confidence?: string | null;

  strengths: string[];
  weaknesses: string[];

  recruiter_insights: InterviewInsight[];

  coaching_tips: string[];

  raw_evaluation: Record<string, unknown>;

  created_at?: string | null;
}

export interface InterviewStartResponse {
  session: InterviewSession;
  first_message: InterviewMessage;
}

export interface InterviewAnswerResponse {
  session: InterviewSession;

  user_message: InterviewMessage;

  ai_message?: InterviewMessage | null;

  result?: InterviewResult | null;
}

export interface InterviewSessionDetailResponse {
  session: InterviewSession;

  messages: InterviewMessage[];

  result?: InterviewResult | null;
}

export interface InterviewStatsBucket {
  difficulty: string;

  sessions: number;

  average_score: number;

  best_score: number;
}

export interface InterviewStatsResponse {
  total_sessions: number;

  average_score: number;

  best_score: number;

  by_difficulty: InterviewStatsBucket[];

  recent_results: InterviewResult[];
}
